packages/tangle/scheduler.go

```diff
@@ -245,6 +266,73 @@ func (s *Scheduler) Clear() {
 	}
 }
 
+// isEligible returns true if the given messageID has either been scheduled or confirmed.
+func (s *Scheduler) isEligible(messageID MessageID) (eligible bool) {
+	s.tangle.Storage.MessageMetadata(messageID).Consume(func(messageMetadata *MessageMetadata) {
+		eligible = messageMetadata.Scheduled() ||
+			s.tangle.ConfirmationOracle.IsMessageConfirmed(messageID) ||
+			messageMetadata.ScheduledBypass() // TO BE REMOVED
+	})
+	return
+}
+
+// tryReady marks the given messageID if all of its parents are eligible and
+// otherwise, returns the list of parents not ready.
+func (s *Scheduler) tryReady(messageID MessageID) (unreadyParents []MessageID) {
+	unreadyParents = s.unreadyParents(messageID)
+	if len(unreadyParents) > 0 {
+		return
+	}
+
+	// all parents are ready
+	if err := s.Ready(messageID); err != nil {
+		s.Events.Error.Trigger(errors.Errorf("failed to mark %s as ready: %w", messageID, err))
+	}
+
+	return
+}
+
+// updateDependencies updates the parentMap with the missing dependencies of the given MessageID
+// and try to set as ready the potential children of the given messageID if no dependencies are given.
+func (s *Scheduler) updateDependencies(messageID MessageID, dependencies []MessageID) {
+	s.parentsMapMutex.Lock()
+	defer s.parentsMapMutex.Unlock()
+
+	if len(dependencies) == 0 {
+		if _, exists := s.parentsMap[messageID]; !exists {
+			return
+		}
+
+		for _, childID := range s.parentsMap[messageID] {
+			s.tryReady(childID)
+		}
+		delete(s.parentsMap, messageID)
```

Comment:
This is the only place where we delete from the map correct?
Is it possible to have a memory overflow attack? 

For example, let say someone sends unsolid messages to the node and overflows the map?
Maybe there are also other conditions in which the map won't be pruned...

Btw when we delete from a map in go, it will actually evict the memory? I recall we had some issues with that...

Reply:
Unsolid messages would not reach the scheduler. This map should always be very small as messages are in general processed already in order. However, if the map should get big, we might want to re-think to use an iteration over the approver instead of using this reverse-map

Reply:
Anyway, I've reverted to use an iteration over the approvers as it seems from the failing tests that the parentsMap was buggy